### ISSUE TITLE
Matchers for child classes aren't working

### ISSUE THREAD

**Issue by Morthan18** [marmer/hamcrest-matcher-generator#32]
Regard following Example:
```java 
// getter and constructors
public class Parent {
	 private final UUID id;
}
// getters and constructors
public class FirstChild extends Parent {
	private final String name;
}
// getter and constructors
public class SecondChild extends Parent {
	private final String name;
}
// getter and constructors
public class Main {
	private final List<Parent> parents;
}
```
I prepared test to check if all Parents that I have in my list, have proper parameters.
```java
MatcherAssert.assertThat(main, isMain()
				.withParents(Matchers.contains(
						isFirstChild()
								.withId(FIRST_ID)
								.withName("First name")),
						isSecondChild()
								.withId(SECOND_ID)
								.withName("Second name")
				))
		);
```
When I run test I'm getting following assertion error:
` property 'parents' item 0: was <FirstChild(super=Parent(id=8dddc07d-3c72-4d43-83b0-a4d8322feb54), name=First name, first=First)`
The problem was occuring when I have some childs with another types.
But if I remove at least one child (no matter which) like below, it works properly.
```java
MatcherAssert.assertThat(main, isMain()
				.withParents(Matchers.hasItem(
						isFirstChild()
								.withId(FIRST_ID)
								.withName("First name"))
				))
		);
```

Currently I solved the problem by extending generated matchers by `TypeSafeMatcher<Parent>`.
I don't want to maintain generated and my modified matcher, so maybe you can fix it easily.

**Comment by marmer** [comment]
Hello Morthan18,

I'm able to reproduce the behavior and I've tried to change the inheritance of the generated child matchers from TypeSafeMatcher<FirstChild> to TypeSafeMatcher<Parent> bit this does not fix it here. Is there something else you've changed to fix this?

Maybe you can share a full working example and I see what I can do to fix this

**Comment by Morthan18** [comment]
I've copied generated matchers to my project package and the only thing I did is change of generic type.
Before:
```java 
@BasedOn(FirstChild.class)
public class FirstChildMatcher extends TypeSafeMatcher<FirstChild> {
    private final BeanPropertyMatcher<FirstChild> beanPropertyMatcher;

    public FirstChildMatcher() {
        beanPropertyMatcher = new BeanPropertyMatcher<FirstChild>(FirstChild.class);
    }

    public FirstChildMatcher withName(final Matcher<?> matcher) {
        beanPropertyMatcher.with("name", matcher);
        return this;
    }

    public FirstChildMatcher withName(final String value) {
        beanPropertyMatcher.with("name", Matchers.equalTo(value));
        return this;
    }

    public FirstChildMatcher withFirst(final Matcher<?> matcher) {
        beanPropertyMatcher.with("first", matcher);
        return this;
    }

    public FirstChildMatcher withFirst(final String value) {
        beanPropertyMatcher.with("first", Matchers.equalTo(value));
        return this;
    }

    public FirstChildMatcher withId(final Matcher<?> matcher) {
        beanPropertyMatcher.with("id", matcher);
        return this;
    }

    public FirstChildMatcher withId(final UUID value) {
        beanPropertyMatcher.with("id", Matchers.equalTo(value));
        return this;
    }

    public FirstChildMatcher withClass(final Matcher<?> matcher) {
        beanPropertyMatcher.with("class", matcher);
        return this;
    }

    public FirstChildMatcher withClass(final Class value) {
        beanPropertyMatcher.with("class", Matchers.equalTo(value));
        return this;
    }

    @Override
    public void describeTo(final Description description) {
        beanPropertyMatcher.describeTo(description);
    }

    @Override
    protected boolean matchesSafely(final FirstChild item) {
        return beanPropertyMatcher.matches(item);
    }

    @Override
    protected void describeMismatchSafely(final FirstChild item, final Description description) {
        beanPropertyMatcher.describeMismatch(item, description);
    }

    public static FirstChildMatcher isFirstChild() {
        return new FirstChildMatcher();
    }
```
After:
```java
@BasedOn(FirstChild.class)
public class FirstChildMatcherCopy extends TypeSafeMatcher<Parent> {
    private final BeanPropertyMatcher<FirstChild> beanPropertyMatcher;

    public FirstChildMatcherCopy() {
        beanPropertyMatcher = new BeanPropertyMatcher<FirstChild>(FirstChild.class);
    }

    public FirstChildMatcherCopy withName(final Matcher<?> matcher) {
        beanPropertyMatcher.with("name", matcher);
        return this;
    }

    public FirstChildMatcherCopy withName(final String value) {
        beanPropertyMatcher.with("name", Matchers.equalTo(value));
        return this;
    }

    public FirstChildMatcherCopy withFirst(final Matcher<?> matcher) {
        beanPropertyMatcher.with("first", matcher);
        return this;
    }

    public FirstChildMatcherCopy withFirst(final String value) {
        beanPropertyMatcher.with("first", Matchers.equalTo(value));
        return this;
    }

    public FirstChildMatcherCopy withId(final Matcher<?> matcher) {
        beanPropertyMatcher.with("id", matcher);
        return this;
    }

    public FirstChildMatcherCopy withId(final UUID value) {
        beanPropertyMatcher.with("id", Matchers.equalTo(value));
        return this;
    }

    public FirstChildMatcherCopy withClass(final Matcher<?> matcher) {
        beanPropertyMatcher.with("class", matcher);
        return this;
    }

    public FirstChildMatcherCopy withClass(final Class value) {
        beanPropertyMatcher.with("class", Matchers.equalTo(value));
        return this;
    }

    @Override
    public void describeTo(final Description description) {
        beanPropertyMatcher.describeTo(description);
    }

    @Override
    protected boolean matchesSafely(final Parent item) {
        return beanPropertyMatcher.matches(item);
    }

    @Override
    protected void describeMismatchSafely(final Parent item, final Description description) {
        beanPropertyMatcher.describeMismatch(item, description);
    }

    public static FirstChildMatcherCopy isFirstChildCopy() {
        return new FirstChildMatcherCopy();
    }
}
```

**Comment by marmer** [comment]
Thank you for the reply. I'm able to reproduce it but I don't jet like the consequences and have to think a little about it.

Because we have to climb up the ladder of inheritance ALL generated matchers could also simply inherit TypeSafeMatcher<Object>. 
Pro: This just works.
Con: The compiler would not be able to help

I Try to find a better solution. If I don't find any, I'll make the change in your suggested way.

**Comment by marmer** [comment]
I just released version 4.2.3. Now all matchers inherit TypeSafeMatcher<Object> and it seams to work properly. So have fun and thanks for reporting this issue and helping to improve this little helper here :)

**Comment by marmer** [comment]
The changes have been reverted in 4.2.4. To keep the help of the compiler with typesafety. As a workaround (so you don't have to modify code) you can simply cast the first argument in contains or hasitems to (Matcher)  or (TypesafeMatcher) as raw type and everything will work fine
